pandas/core/indexes/datetimes.py

```diff
@@ -191,6 +191,7 @@ class DatetimeIndex(DatetimeTimedeltaMixin):
 
     _data: DatetimeArray
     tz: Optional[tzinfo]
+    tz_localize: Callable[..., "DatetimeIndex"]
```

Comment:
I try to fake-annotate [this tz_localize implementation](https://github.com/pandas-dev/pandas/blob/master/pandas/core/arrays/datetimes.py#L822).
I tried using something like `Callable[["DatetimeIndex", Optional[str], str, str], "DatetimeIndex"]`, but then mypy would raise errors when not omitting some arguments like `index.tz_localize("US/Eastern")`

Reply:
The original problem was that `DatetimeIndex(["2011-01-01", "2011-01-02"]).tz_localize("US/Eastern")` [here](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/arithmetic/test_datetime64.py#L2090) would raise:

`"DatetimeIndex" has no attribute "tz_localize"`

Reply:
this is going to be a common problem with `inherit_names`.  best case would be to find a way to fix it at the source

Reply:
What is the original mypy error? I'd be hesitant to make any changes to the actual code base to silence errors coming from the test suite

Reply:
@WillAyd I stated it in my second comment. I think @jbrockmendel is right, but it's not clear how a solution might look like. I'm investigating though ✌️

Reply:
After pondering on this for a while, I still have no clue on how to tackle this issue... As long as we assign these methods dynamically instead of using some sort of inheritance, I don't think mypy offers a feature to support that.

If anyone can point me in a direction, I'd be happy to explore it further though.

@WillAyd Regarding your concern on changing the actual code. That's not really the case. It's just adding a type annotation. These will be ignored when running the python code; just mypy takes them into account. Also, I'm sure we'll run into the same issue when annotating more of the "actual" codebase - so it's better to think about solutions now.
As @jbrockmendel pointed out, the biggest problem will be scalability: With my proposed solution we'd basically need to add very broad type annotations for all dynamically assigned methods...

Reply:
> As long as we assign these methods dynamically instead of using some sort of inheritance, I don't think mypy offers a feature to support that.

correct. mypy is a static type checker. so either the inherit_names decorator needs to be replaced with a mixin in order for mypy to be able to perform static checks or mypy needs to be given additional information in the form of type declarations.

I recommended the type declaration approach in the original PR, https://github.com/pandas-dev/pandas/pull/29041/files#r345832865

Reply:
So that's what I did, right? Or do you have something else in mind?

Reply:
yeah, I'm happy with this, but it seems the other reviewers may have an alternative.

Reply:
Yea I still am not a huge fan of this design - is the plan to just keep patching these individually for every dynamically bound attribute?

Reply:
@WillAyd Should I create a ticket to have a broader discussion on the subject? Seems like we need to agree on a more general approach.

Reply:
Do you see any alternatives? Again my concern is that going in this direction will yield a lot of code duplication

Reply:
@WillAyd Like mentioned above: No. As pointed out by simonjayhawkins, mypy is a static type checker. So by design, mypy cannot understand the dynamic behaviour of `inherit_names`.

The only alternative would be to refactor `inherit_names` to work by inheritance instead of dynamic assignment. That would be a separate issue though.

Reply:
Yea so to be explicit I am -1 on this approach. Would rather just ignore for now until a more comprehensive solution can be found

Reply:
It seems like this would be the last issue where an annotation like this is required to finish #28926 off. Therefore it's maybe worth still merging this one and working on a more general solution afterwards? Wdyt @WillAyd @simonjayhawkins?

Reply:
I would prefer to either ignore where required in a few places or just leave things as they are today

Reply:
Alright, I'll close this PR then and create a follow-up issue that focuses on refactoring `inherit_names`

Reply:
xref: #32100

Reply:
Great thanks for understanding - your contributions are very much appreciated